### PR TITLE
Fix collect clause not supported inside a query action statement

### LINKED ISSUE #41392: [Bug]: Compilation fails when do clause is present after an aggregation
### Description

Below ballerina code is a syntactically and semantically correct. But when we compile it gives compile-time errors.
```ballerina
function total(float[][] items) {
    from var i in items
    let float tot = from var j in i collect sum(j)
    do {
        
    };
}
```

Compile time errors are,
```
invalid expression statement
missing select clause
incompatible types: expected 'float', found '()'
```

### Steps to Reproduce

_No response_

### Affected Version(s)

2201.8.0-RC1

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

Fixes #41392 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query-action.bal [hunk 1]
```diff
@@ -862,6 +862,17 @@ function testJoinedQueryActionWithRegExp() {
     "(?:A|B|[ab-fgh]+(?im-x:[cdeg-k]??AAAB*[^abc-efg](?:A|B|[ab-fgh]+(?im-x:[cdeg-k]??A)|)|^|PQ?)|)|^|PQ?");
 }
 
+function testQueryActionWithCollectClauseInsideLeClause() {
```
Comment (SasinduDilshara) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/query-action.bal:865]: Shall we add one test with similar to the issue,
with an empty do clause.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -11949,7 +11949,8 @@ private STNode parseQueryExprRhs(STNode queryConstructType, boolean isRhsExpr, b
             }
         }
 
-        if (peek().kind == SyntaxKind.DO_KEYWORD && (!isNestedQueryExpr() || selectClause == null)) {
+        if (peek().kind == SyntaxKind.DO_KEYWORD &&
```
Comment (pcnfernando) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:11952]: Shall we add few parser tests for this. include some with collect clause inside do, group-by clauses

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query-action.bal (hunk 1 above)
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/query-action.bal]: Added in commit [#f0b8530](https://github.com/ballerina-platform/ballerina-lang/pull/41413/commits/f0b85305ea284059a10559877e20523b3a47cf94)

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java (hunk 2 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:11952]: Added in commit [#6edfe50](https://github.com/ballerina-platform/ballerina-lang/pull/41413/commits/6edfe50fb322850be8689b622fe8dc2e00155b1a)

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query-action.bal [hunk 3]
```diff
@@ -862,6 +862,26 @@ function testJoinedQueryActionWithRegExp() {
...
+    int total = 0;
+    from var i in arr
+        let int tot = from var j in i collect sum(j)
+        do {
+            total += tot;
+        };
+    assertEquality(21, total);
+}
+
+function testQueryActionWithCollectClauseInsideLeClause2() {
```
Comment (SasinduDilshara) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/query-action.bal:876]: The original issue intended when we passing float array as a functional argument, Did you check that scenario with tests?

Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/query-action.bal:876]: The problem resides within the parser, and the choice of the collection in the query is unrelated.

**Review by SasinduDilshara — DISMISSED** [review]

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 4]
```diff
@@ -11949,7 +11949,8 @@ private STNode parseQueryExprRhs(STNode queryConstructType, boolean isRhsExpr, b
             }
         }
 
-        if (peek().kind == SyntaxKind.DO_KEYWORD && (!isNestedQueryExpr() || selectClause == null)) {
+        if (peek().kind == SyntaxKind.DO_KEYWORD && 
+                (!isNestedQueryExpr() || (selectClause == null && collectClause == null))) {
             STNode intermediateClauses = STNodeFactory.createNodeList(clauses);
             STNode queryPipeline = STNodeFactory.createQueryPipelineNode(fromClause, intermediateClauses);
             return parseQueryAction(queryConstructType, queryPipeline, selectClause);
```
Comment (KavinduZoysa) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:11956]: We have to pass `collectClause` here and handle it.

**Review by KavinduZoysa — APPROVED** [review]

File: compiler/ballerina-parser/src/test/resources/test_parser.bal [hunk 5]
```diff
@@ -1,2 +1,14 @@
-public function main() {
+function foo() {
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/test/resources/test_parser.bal:1]: @LakshanWeerasinghe Do we need to change this file? Shall we change this back to the original, please?

Comment (LakshanWeerasinghe) [reply compiler/ballerina-parser/src/test/resources/test_parser.bal:1]: Fix in #42254.

**Pull request merged**